File: frontend-next/lib/os-api/reg45-quality-review.ts
'use client'

export type Reg45ReviewStatus =
  | 'draft'
  | 'evidence_gathering'
  | 'ready_for_manager_review'
  | 'manager_reviewed'
  | 'ri_review_required'
  | 'ri_reviewed'
  | 'finalised'
  | 'archived'

export type Reg45EvidenceStrength =
  | 'strong_evidence'
  | 'partial_evidence'
  | 'draft_only'
  | 'prompt_only'
  | 'route_hint_only'
  | 'not_yet_wired'
  | 'not_safe_to_summarise'

export type Reg45EvidenceRisk = 'low' | 'medium' | 'high' | 'urgent'

export type Reg45ReviewEvidenceItem = {
  id: string
  title: string
  safe_summary: string
  source_module: string
  route: string
  action_label?: string | null
  section_types?: string[]
  evidence_strength: Reg45EvidenceStrength
  risk: Reg45EvidenceRisk
  manager_review_required?: boolean
  safeguarding_review_required?: boolean
  metadata?: Record<string, unknown>
}

export type Reg45ReviewGap = {
  id: string
  title: string
  description: string
  section_type: string
  risk: Reg45EvidenceRisk
  route: string
  action_label?: string | null
  recommended_action: string
}

export type Reg45ImprovementActionDraft = {
  id: string
  title: string
  description: string
  source_gap_id?: string | null
  priority: string
  route: string
}

export type Reg45ReviewSection = {
  id: string
  title: string
  section_type: string
  summary: string
  evidence_items: Reg45ReviewEvidenceItem[]
  gaps: Reg45ReviewGap[]
  improvement_actions: Reg45ImprovementActionDraft[]
  warnings: string[]
}

export type Reg45QualityReview = {
  id: string
  title: string
  status: Reg45ReviewStatus
  generated_at: string
  period_start?: string | null
  period_end?: string | null
  summary: string
  sections: Reg45ReviewSection[]
  evidence_count: number
  gap_count: number
  draft_only_count: number
  improvement_action_count: number
  review_required_count: number
  safeguarding_review_count: number
  limitations: string[]
  privacy_notice: string
  orb_prompts: Array<{ label: string; mode: string; query: string }>
  routes: Record<string, string>
  metadata?: Record<string, unknown>
}

export type Reg45ReviewDashboard = {
  generated_at: string
  summary: string
  draft_review_count: number
  ready_for_manager_count: number
  ri_review_required_count: number
  recent_reviews: Array<Record<string, unknown>>
  key_gaps: Reg45ReviewGap[]
  recommendations: string[]
  privacy_notice: string
  routes: Record<string, string>
}

type ApiEnvelope<T> = { success?: boolean; data?: T; error?: string }

async function parseEnvelope<T>(response: Response, fallback: T): Promise<{ data: T; ok: boolean; error?: string }> {
  if (!response.ok) {
    return { data: fallback, ok: false, error: `HTTP ${response.status}` }
  }
  const payload = (await response.json().catch(() => ({}))) as ApiEnvelope<T> | T
  const envelope = payload as ApiEnvelope<T>
  return {
    data: envelope && typeof envelope === 'object' && 'data' in envelope ? (envelope.data as T) : (payload as T),
    ok: true
  }
}

export function reg45ReviewOrbHref(mode: string, query: string): string {
  const params = new URLSearchParams({ mode, q: query })
  return `/assistant/orb?${params.toString()}`
}

export async function getReg45Health() {
  const response = await fetch('/api/reg45/health', { credentials: 'include', cache: 'no-store' })
  return parseEnvelope(response, { status: 'unavailable' })
}

export async function getReg45Dashboard() {
  const response = await fetch('/api/reg45/dashboard', { credentials: 'include', cache: 'no-store' })
  const empty: Reg45ReviewDashboard = {
    generated_at: '',
    summary: '',
    draft_review_count: 0,
    ready_for_manager_count: 0,
    ri_review_required_count: 0,
    recent_reviews: [],
    key_gaps: [],
    recommendations: [],
    privacy_notice: '',
    routes: {}
  }
  return parseEnvelope(response, empty)
}

export async function generateReg45Review(options?: {
  title?: string
  period_start?: string
  period_end?: string
  from_inspection_pack_id?: string
  save_draft?: boolean
  create_improvement_actions?: boolean
}) {
  const response = await fetch('/api/reg45/reviews/generate', {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      save_draft: true,
      create_improvement_actions: false,
      ...options
    })
  })
  const empty: Reg45QualityReview = {
    id: '',
    title: '',
    status: 'draft',
    generated_at: '',
    summary: '',
    sections: [],
    evidence_count: 0,
    gap_count: 0,
    draft_only_count: 0,
    improvement_action_count: 0,
    review_required_count: 0,
    safeguarding_review_count: 0,
    limitations: [],
    privacy_notice: '',
    orb_prompts: [],
    routes: {}
  }
  return parseEnvelope(response, empty)
}

export async function listReg45Reviews(limit = 20) {
  const response = await fetch(`/api/reg45/reviews?limit=${limit}`, {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope(response, { reviews: [] as Array<Record<string, unknown>> })
}

export async function getReg45Review(reviewId: string) {
  const response = await fetch(`/api/reg45/reviews/${encodeURIComponent(reviewId)}`, {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope(response, {} as Reg45QualityReview)
}

export async function updateReg45Review(
  reviewId: string,
  body: { status?: Reg45ReviewStatus; reviewer_notes?: string; metadata?: Record<string, unknown> }
) {
  const response = await fetch(`/api/reg45/reviews/${encodeURIComponent(reviewId)}`, {
    method: 'PATCH',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  })
  return parseEnvelope(response, {} as Reg45QualityReview)
}

export async function applyReg45ReviewAction(
  reviewId: string,
  action: string,
  note?: string
) {
  const response = await fetch(`/api/reg45/reviews/${encodeURIComponent(reviewId)}/action`, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ action, note })
  })
  return parseEnvelope(response, { success: false, warnings: [] })
}

export async function exportReg45Review(reviewId: string) {
  const response = await fetch(`/api/reg45/reviews/${encodeURIComponent(reviewId)}/export`, {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope(response, { markdown: '', disclaimer: '' })
}

export async function createReg45ActionsFromGaps(reviewId: string) {
  const response = await fetch(`/api/reg45/reviews/${encodeURIComponent(reviewId)}/create-actions`, {
    method: 'POST',
    credentials: 'include'
  })
  return parseEnvelope(response, { action_ids: [] as string[], warning: null as string | null })
}

export function exportReviewMarkdownClient(review: Reg45QualityReview): string {
  const lines = [`# ${review.title}`, '', review.summary, '']
  for (const section of review.sections) {
    lines.push(`## ${section.title}`, section.summary, '')
    for (const item of section.evidence_items) {
      lines.push(`- ${item.title}: ${item.safe_summary}`)
    }
    for (const gap of section.gaps) {
      lines.push(`- Potential gap: ${gap.title}`)
    }
  }
  lines.push('', 'Draft review — requires manager/provider review.')
  return lines.join('\n')
}

File: schemas/reg45_quality_review.py
"""Reg 45 Quality of Care Review schemas — draft review support, metadata only."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from schemas.sccif_alignment import QualityStandardArea, SccifJudgementArea

Reg45ReviewStatus = Literal[
    "draft",
    "evidence_gathering",
    "ready_for_manager_review",
    "manager_reviewed",
    "ri_review_required",
    "ri_reviewed",
    "finalised",
    "archived",
]

Reg45ReviewSectionType = Literal[
    "summary",
    "child_voice",
    "progress_outcomes",
    "quality_purpose",
    "safeguarding_protection",
    "education",
    "health_wellbeing",
    "positive_relationships",
    "care_planning",
    "workforce_leadership",
    "patterns_themes",
    "improvement_actions",
    "provider_ri_review",
    "final_reflections",
]

Reg45EvidenceStrength = Literal[
    "strong_evidence",
    "partial_evidence",
    "draft_only",
    "prompt_only",
    "route_hint_only",
    "not_yet_wired",
    "not_safe_to_summarise",
]

Reg45EvidenceRisk = Literal["low", "medium", "high", "urgent"]


class Reg45OfficialSourceRef(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    title: str
    url: str | None = None
    note: str | None = None


class Reg45ReviewEvidenceItem(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    title: str
    safe_summary: str
    source_module: str
    source_type: str = "metadata"
    route: str
    action_label: str | None = None
    section_types: list[Reg45ReviewSectionType] = Field(default_factory=list)
    quality_standards: list[QualityStandardArea] = Field(default_factory=list)
    sccif_judgement_areas: list[SccifJudgementArea] = Field(default_factory=list)
    evidence_strength: Reg45EvidenceStrength = "partial_evidence"
    draft_status: str | None = None
    risk: Reg45EvidenceRisk = "low"
    review_required: bool = False
    manager_review_required: bool = False
    safeguarding_review_required: bool = False
    privacy_sensitive: bool = False
    related_id: str | None = None
    related_type: str | None = None
    child_id: int | None = None
    staff_id: str | None = None
    home_id: int | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class Reg45ReviewGap(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    title: str
    description: str
    section_type: Reg45ReviewSectionType
    quality_standard: QualityStandardArea | None = None
    risk: Reg45EvidenceRisk = "medium"
    route: str
    action_label: str | None = None
    recommended_action: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)


class Reg45ImprovementActionDraft(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    title: str
    description: str
    source_gap_id: str | None = None
    priority: Literal["low", "medium", "high", "urgent"] = "medium"
    suggested_owner_role: str = "registered_manager"
    due_in_days: int | None = None
    route: str = "/actions"
    action_label: str | None = "Review action"
    metadata: dict[str, Any] = Field(default_factory=dict)


class Reg45ReviewSection(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    title: str
    section_type: Reg45ReviewSectionType
    summary: str
    evidence_items: list[Reg45ReviewEvidenceItem] = Field(default_factory=list)
    gaps: list[Reg45ReviewGap] = Field(default_factory=list)
    improvement_actions: list[Reg45ImprovementActionDraft] = Field(default_factory=list)
    reviewer_notes: str | None = None
    warnings: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class Reg45QualityReview(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    title: str
    status: Reg45ReviewStatus = "draft"
    generated_at: str
    period_start: str | None = None
    period_end: str | None = None
    home_id: int | None = None
    summary: str
    sections: list[Reg45ReviewSection] = Field(default_factory=list)
    evidence_count: int = 0
    gap_count: int = 0
    draft_only_count: int = 0
    improvement_action_count: int = 0
    review_required_count: int = 0
    safeguarding_review_count: int = 0
    limitations: list[str] = Field(default_factory=list)
    privacy_notice: str = ""
    official_sources: list[Reg45OfficialSourceRef] = Field(default_factory=list)
    orb_prompts: list[dict[str, str]] = Field(default_factory=list)
    routes: dict[str, str] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


class Reg45ReviewCreateRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    title: str | None = None
    period_start: str | None = None
    period_end: str | None = None
    home_id: int | None = None
    from_inspection_pack_id: str | None = None
    save_draft: bool = True
    create_improvement_actions: bool = False


class Reg45ReviewUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    status: Reg45ReviewStatus | None = None
    reviewer_notes: str | None = None
    sections: list[Reg45ReviewSection] | None = None
    metadata: dict[str, Any] | None = None


Reg45ReviewActionType = Literal[
    "mark_ready_for_manager_review",
    "mark_manager_reviewed",
    "request_ri_review",
    "mark_ri_reviewed",
    "finalise",
    "archive",
    "create_actions_from_gaps",
]


class Reg45ReviewActionRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    action: Reg45ReviewActionType
    note: str | None = None
    metadata: dict[str, Any] | None = None


class Reg45ReviewActionResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    success: bool = True
    review: Reg45QualityReview | None = None
    previous_status: str | None = None
    new_status: str | None = None
    action_ids: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    message: str = ""


class Reg45ReviewDashboard(BaseModel):
    model_config = ConfigDict(extra="ignore")

    generated_at: str
    summary: str
    draft_review_count: int = 0
    ready_for_manager_count: int = 0
    ri_review_required_count: int = 0
    recent_reviews: list[dict[str, Any]] = Field(default_factory=list)
    key_gaps: list[Reg45ReviewGap] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)
    limitations: list[str] = Field(default_factory=list)
    privacy_notice: str = ""
    routes: dict[str, str] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


class Reg45ReviewHealth(BaseModel):
    model_config = ConfigDict(extra="ignore")

    status: str = "ok"
    persistence_available: bool = False
    sources_available: list[str] = Field(default_factory=list)
    limitations: list[str] = Field(default_factory=list)
